File: protocol-generator/haskell/src/EntityCore/TypeDefs.hs
{-# LANGUAGE OverloadedStrings #-}

-- | Core type floor (V7 §9.5) — the render-from-model registry (A-HS-009 resolved
-- at S4).
--
-- Design (cross-peer ruling): peers render @system/type/*@ from an in-code model
-- through the native S2 codec, NOT by ingesting fixture bytes — the entity's
-- content_hash is computed by our own encoder over the model, then byte-diffed
-- against the canonical type-registry vectors (the S8 drift target). A core peer
-- publishes exactly the §9.5 floor — the 53 core + operational + type-system
-- bootstrap types; extension vocabularies (compute\/*, content\/*, …) are NOT
-- published by a core peer (refined G4 \/ F17). The oracle's @type_system@
-- category matches the 53 floor as a hard FAIL gate and WARNs (matched-if-present)
-- on the non-floor types it also probes.
--
-- Omit-empty: an absent\/false\/zero field drops the key, so the rendered ECF map
-- is byte-identical to the Go reference encoder. The codec sorts map keys
-- canonically (length-then-lex), so field declaration order is irrelevant to the
-- bytes — only the present key\/value set matters.
--
-- Faithful port of the cross-blessed C#\/TS\/OCaml\/Zig registry (byte-identical
-- to the Go oracle vectors).
module EntityCore.TypeDefs
  ( coreTypes
  , coreTypeCount
  , publish
    -- * Render-from-model internals (exported for the S4 byte-diff test)
  , FSpec (..)
  , TypeDef (..)
  , fref
  , opt
  , sized
  , farray
  , fmapOf
  , funion
  , typeDefData
  , allTypes
  ) where

import Data.Text (Text)

import EntityCore.Codec.Value (Value (..))
import EntityCore.Model (Entity, makeEntity)
import EntityCore.Store (Store, bind)

-- ── FSpec — a field spec inside a TypeDef (system/type/field-spec shape) ──────
--
-- Exactly one structural carrier is set: a type_ref, an array_of, a map_of, or a
-- union_of. Rendered omit-empty into the field-spec ECF map.

data FSpec = FSpec
  { fsTypeRef  :: Maybe Text
  , fsOptional :: Bool
  , fsArrayOf  :: Maybe FSpec
  , fsMapOf    :: Maybe FSpec
  , fsUnionOf  :: Maybe [FSpec]
  , fsKeyType  :: Maybe Text
  , fsByteSize :: Maybe Word
  }

emptyFSpec :: FSpec
emptyFSpec = FSpec Nothing False Nothing Nothing Nothing Nothing Nothing

-- | @type_ref@ to a named type.
fref :: Text -> FSpec
fref t = emptyFSpec { fsTypeRef = Just t }

-- | Mark a field-spec optional.
opt :: FSpec -> FSpec
opt s = s { fsOptional = True }

-- | Add a @byte_size@ carrier.
sized :: Word -> FSpec -> FSpec
sized n s = s { fsByteSize = Just n }

-- | @array_of@ the given element spec.
farray :: FSpec -> FSpec
farray elemSpec = emptyFSpec { fsArrayOf = Just elemSpec }

-- | @map_of@ the given value spec, with an optional @key_type@.
fmapOf :: Maybe Text -> FSpec -> FSpec
fmapOf keyT valSpec = emptyFSpec { fsMapOf = Just valSpec, fsKeyType = keyT }

-- | @union_of@ the given variant specs.
funion :: [FSpec] -> FSpec
funion vs = emptyFSpec { fsUnionOf = Just vs }

-- | Render a field-spec to its ECF data map (omit-empty).
fspecData :: FSpec -> Value
fspecData s =
  VMap $ concat
    [ kv "type_ref" VText (fsTypeRef s)
    , [ (VText "optional", VBool True) | fsOptional s ]
    , kv "array_of" fspecData (fsArrayOf s)
    , kv "map_of" fspecData (fsMapOf s)
    , case fsUnionOf s of
        Just vs -> [ (VText "union_of", VArray (map fspecData vs)) ]
        Nothing -> []
    , kv "key_type" VText (fsKeyType s)
    , kv "byte_size" (VUInt . fromIntegral) (fsByteSize s)
    ]
  where
    kv :: Text -> (a -> Value) -> Maybe a -> [(Value, Value)]
    kv k f = maybe [] (\x -> [(VText k, f x)])

-- ── TypeDef — a core type definition (system/type entity data) ────────────────

data TypeDef = TypeDef
  { tdName    :: Text
  , tdExtends :: Maybe Text
  , tdFields  :: [(Text, FSpec)]
  , tdLayout  :: [Text]
  }

def :: Text -> TypeDef
def n = TypeDef n Nothing [] []

-- These builders are used infix (@td \`withFields\` fields@) so the @TypeDef@ is
-- the LEFT operand — hence @td@ comes first.
withExtends :: TypeDef -> Text -> TypeDef
withExtends td e = td { tdExtends = Just e }

withFields :: TypeDef -> [(Text, FSpec)] -> TypeDef
withFields td fs = td { tdFields = fs }

withLayout :: TypeDef -> [Text] -> TypeDef
withLayout td ls = td { tdLayout = ls }

-- | Render the @system/type@ data map (omit-empty). Field declaration order is
-- preserved within the @fields@ sub-map; the codec re-sorts keys canonically.
typeDefData :: TypeDef -> Value
typeDefData td =
  VMap $ concat
    [ [ (VText "name", VText (tdName td)) ]
    , maybe [] (\e -> [(VText "extends", VText e)]) (tdExtends td)
    , [ (VText "fields", VMap [ (VText k, fspecData s) | (k, s) <- tdFields td ])
      | not (null (tdFields td)) ]
    , [ (VText "layout", VArray (map VText (tdLayout td)))
      | not (null (tdLayout td)) ]
    ]

typeDefEntity :: TypeDef -> Entity
typeDefEntity td = makeEntity "system/type" (typeDefData td)

-- ── reused nested specs ───────────────────────────────────────────────────────

spString, spAny, spHash, spCoreEntity, spTreePath, spGrantEntry, spMultiGranter :: FSpec
spString       = fref "primitive/string"
spAny          = fref "primitive/any"
spHash         = fref "system/hash"
spCoreEntity   = fref "core/entity"
spTreePath     = fref "system/tree/path"
spGrantEntry   = fref "system/capability/grant-entry"
spMultiGranter = fref "system/capability/multi-granter"

spFieldSpec, spOpSpec, spListingEntry, spType, spTypeName :: FSpec
spFieldSpec    = fref "system/type/field-spec"
spOpSpec       = fref "system/handler/operation-spec"
spListingEntry = fref "system/tree/listing-entry"
spType         = fref "system/type"
spTypeName     = fref "system/type/name"

-- ── the 53 core type definitions ──────────────────────────────────────────────

allTypes :: [TypeDef]
allTypes =
  [ -- primitives (8)
    def "primitive/any"
  , def "primitive/bool"
  , def "primitive/bytes"
  , def "primitive/float"
  , def "primitive/int"
  , def "primitive/null"
  , def "primitive/string"
  , def "primitive/uint"

    -- structural roots + envelopes (5)
  , def "entity" `withFields`
      [ ("type", fref "primitive/string")
      , ("data", fref "primitive/any")
      ]
  , def "core/entity" `withFields`
      [ ("type", fref "primitive/string")
      , ("data", fref "primitive/any")
      , ("content_hash", fref "system/hash")
      ]
  , def "core/envelope" `withFields`
      [ ("root", fref "core/entity")
      , ("included", opt (fmapOf (Just "system/hash") spCoreEntity))
      ]
  , def "system/envelope" `withExtends` "core/envelope"
  , def "system/protocol/envelope" `withExtends` "core/envelope"

    -- identity / hash / signature (4)
  , (def "system/hash" `withExtends` "primitive/bytes") `withFields`
      [ ("format_code", sized 1 (fref "primitive/uint"))
      , ("digest", fref "primitive/bytes")
      ] `withLayout` ["format_code", "digest"]
  , def "system/peer" `withFields`
      [ ("key_type", fref "primitive/string")
      , ("peer_id", fref "system/peer-id")
      , ("public_key", fref "primitive/bytes")
      ]
  , def "system/peer-id" `withExtends` "primitive/string"
  , def "system/signature" `withFields`
      [ ("algorithm", fref "primitive/string")
      , ("signature", fref "primitive/bytes")
      , ("signer", fref "system/hash")
      , ("target", fref "system/hash")
      ]

    -- protocol surface (6)
  , def "system/protocol/connect/authenticate" `withFields`
      [ ("key_type", fref "primitive/string")
      , ("nonce", fref "primitive/bytes")
      , ("peer_id", fref "system/peer-id")
      , ("public_key", fref "primitive/bytes")
      ]
  , def "system/protocol/connect/hello" `withFields`
      [ ("protocols", farray spString)
      , ("nonce", fref "primitive/bytes")
      , ("peer_id", fref "system/peer-id")
      , ("timestamp", fref "primitive/uint")
      , ("compression", opt (farray spString))
      , ("encryption", opt (farray spString))
      , ("hash_formats", opt (farray spString))
      , ("key_types", opt (farray spString))
      ]
  , def "system/protocol/error" `withFields`
      [ ("code", fref "primitive/string")
      , ("message", opt (fref "primitive/string"))
      , ("rejected_marker", opt (fref "system/hash"))
      ]
  , def "system/protocol/execute" `withFields`
      [ ("operation", fref "primitive/string")
      , ("params", fref "core/entity")
      , ("request_id", fref "primitive/string")
      , ("uri", fref "system/tree/path")
      , ("author", opt (fref "system/hash"))
      , ("bounds", opt (fref "system/bounds"))
      , ("capability", opt (fref "system/hash"))
      , ("deliver_to", opt (fref "system/delivery-spec"))
      , ("deliver_token", opt (fref "system/hash"))
      , ("durability_request", opt (fref "system/durability-request"))
      , ("resource", opt (fref "system/protocol/resource-target"))
      ]
  , def "system/protocol/execute/response" `withFields`
      [ ("request_id", fref "primitive/string")
      , ("result", fref "core/entity")
      , ("status", fref "primitive/uint")
      , ("durability", opt (fref "system/durability-result"))
      ]
  , def "system/protocol/resource-target" `withFields`
      [ ("targets", farray spTreePath)
      , ("exclude", opt (farray spTreePath))
      ]

    -- capability (12)
  , def "system/capability/grant" `withFields`
      [ ("token", fref "system/hash") ]
  , def "system/capability/grant-entry" `withFields`
      [ ("handlers", fref "system/capability/path-scope")
      , ("operations", fref "system/capability/id-scope")
      , ("resources", fref "system/capability/path-scope")
      , ("allowances", opt (fmapOf Nothing spAny))
      , ("constraints", opt (fmapOf Nothing spAny))
      , ("peers", opt (fref "system/capability/id-scope"))
      ]
  , def "system/capability/id-scope" `withFields`
      [ ("include", farray spString)
      , ("exclude", opt (farray spString))
      ]
  , def "system/capability/path-scope" `withFields`
      [ ("include", farray spTreePath)
      , ("exclude", opt (farray spTreePath))
      ]
  , def "system/capability/request" `withFields`
      [ ("grants", farray spGrantEntry)
      , ("ttl_ms", opt (fref "primitive/uint"))
      ]
  , def "system/capability/revocation" `withFields`
      [ ("token", fref "system/hash")
      , ("revoked_at", fref "primitive/uint")
      , ("reason", opt (fref "primitive/string"))
      ]
  , def "system/capability/revoke-request" `withFields`
      [ ("token", fref "system/hash")
      , ("reason", opt (fref "primitive/string"))
      ]
  , def "system/capability/delegate-request" `withFields`
      [ ("grants", farray spGrantEntry)
      , ("parent", fref "system/hash")
      , ("ttl_ms", opt (fref "primitive/uint"))
      ]
  , def "system/capability/delegation-caveats" `withFields`
      [ ("max_delegation_depth", opt (fref "primitive/uint"))
      , ("max_delegation_ttl", opt (fref "primitive/uint"))
      , ("no_delegation", opt (fref "primitive/bool"))
      ]
  , def "system/capability/policy-entry" `withFields`
      [ ("grants", farray spGrantEntry)
      , ("peer_pattern", fref "primitive/string")
      , ("notes", opt (fref "primitive/string"))
      , ("ttl_ms", opt (fref "primitive/uint"))
      ]
  , def "system/capability/token" `withFields`
      [ ("created_at", fref "primitive/uint")
      , ("grantee", fref "system/hash")
      , ("granter", funion [spHash, spMultiGranter])
      , ("grants", farray spGrantEntry)
      , ("delegation_caveats", opt (fref "system/capability/delegation-caveats"))
      , ("expires_at", opt (fref "primitive/uint"))
      , ("not_before", opt (fref "primitive/uint"))
      , ("parent", opt (fref "system/hash"))
      , ("resource_limits", opt (fref "system/resource-limits"))
      ]
  , def "system/capability/multi-granter" `withFields`
      [ ("signers", farray spHash)
      , ("threshold", fref "primitive/uint")
      ]

    -- handler machinery (6)
  , def "system/handler" `withFields`
      [ ("interface", fref "system/tree/path")
      , ("expression_path", opt (fref "system/tree/path"))
      , ("internal_scope", opt (farray spGrantEntry))
      , ("max_scope", opt (farray spGrantEntry))
      ]
  , def "system/handler/interface" `withFields`
      [ ("name", fref "primitive/string")
      , ("operations", fmapOf Nothing spOpSpec)
      , ("pattern", fref "system/tree/path")
      ]
  , (def "system/handler/manifest" `withExtends` "system/handler/interface") `withFields`
      [ ("name", fref "primitive/string")
      , ("operations", fmapOf Nothing spOpSpec)
      , ("pattern", fref "system/tree/path")
      , ("expression_path", opt (fref "system/tree/path"))
      , ("internal_scope", opt (farray spGrantEntry))
      , ("max_scope", opt (farray spGrantEntry))
      ]
  , def "system/handler/operation-spec" `withFields`
      [ ("input_type", opt (fref "system/type/name"))
      , ("output_type", opt (fref "system/type/name"))
      ]
  , def "system/handler/register-request" `withFields`
      [ ("manifest", fref "system/handler/manifest")
      , ("requested_scope", opt (farray spGrantEntry))
      , ("types", opt (fmapOf Nothing spType))
      ]
  , def "system/handler/register-result" `withFields`
      [ ("grant", fref "system/capability/token")
      , ("pattern", fref "system/tree/path")
      ]

    -- tree (5)
  , def "system/tree/get-request" `withFields`
      [ ("limit", opt (fref "primitive/uint"))
      , ("mode", opt (fref "primitive/string"))
      , ("offset", opt (fref "primitive/uint"))
      , ("tree_id", opt (fref "primitive/string"))
      ]
  , def "system/tree/put-request" `withFields`
      [ ("entity", opt (fref "core/entity"))
      , ("expected_hash", opt (fref "system/hash"))
      , ("tree_id", opt (fref "primitive/string"))
      ]
  , def "system/tree/listing" `withFields`
      [ ("count", fref "primitive/uint")
      , ("entries", fmapOf Nothing spListingEntry)
      , ("offset", fref "primitive/uint")
      , ("path", fref "system/tree/path")
      , ("next_page", opt (fref "system/hash"))
      ]
  , def "system/tree/listing-entry" `withFields`
      [ ("has_children", fref "primitive/bool")
      , ("hash", opt (fref "system/hash"))
      ]
  , def "system/tree/path" `withExtends` "primitive/string"

    -- type-system bootstrap (3)
  , def "system/type" `withFields`
      [ ("name", fref "system/type/name")
      , ("extends", opt (fref "system/type/name"))
      , ("fields", opt (fmapOf Nothing spFieldSpec))
      , ("layout", opt (farray spString))
      , ("type_args", opt (fmapOf Nothing spTypeName))
      , ("type_params", opt (farray spString))
      ]
  , def "system/type/field-spec" `withFields`
      [ ("type_ref", opt (fref "system/type/name"))
      , ("optional", opt (fref "primitive/bool"))
      , ("array_of", opt (fref "system/type/field-spec"))
      , ("map_of", opt (fref "system/type/field-spec"))
      , ("union_of", opt (farray spFieldSpec))
      , ("key_type", opt (fref "system/type/name"))
      , ("byte_size", opt (fref "primitive/uint"))
      , ("type_param", opt (fref "primitive/string"))
      , ("type_args", opt (fmapOf Nothing spTypeName))
      , ("default", opt (fref "primitive/any"))
      , ("constraints", opt (farray spCoreEntity))
      ]
  , def "system/type/name" `withExtends` "primitive/string"

    -- operational (4)
  , def "system/bounds" `withFields`
      [ ("budget", opt (fref "primitive/uint"))
      , ("cascade_depth", opt (fref "primitive/uint"))
      , ("chain_id", opt (fref "primitive/string"))
      , ("parent_chain_id", opt (fref "primitive/string"))
      , ("ttl", opt (fref "primitive/uint"))
      , ("visited", opt (farray spTreePath))
      ]
  , def "system/resource-limits" `withFields`
      [ ("max_budget", opt (fref "primitive/uint"))
      , ("max_ttl", opt (fref "primitive/uint"))
      , ("max_visited_length", opt (fref "primitive/uint"))
      ]
  , def "system/delivery-spec" `withFields`
      [ ("operation", fref "primitive/string")
      , ("uri", fref "system/tree/path")
      ]
  , def "system/deletion-marker"
  ]

-- | The number of core types published (53).
coreTypeCount :: Int
coreTypeCount = length allTypes

-- | The rendered @system/type@ entities for all core types, keyed by name.
coreTypes :: [(Text, Entity)]
coreTypes = [ (tdName td, typeDefEntity td) | td <- allTypes ]

-- | Publish every core type at @\/{peer}\/system\/type\/{name}@ (content-addressed
-- in the store too, via 'bind').
publish :: Store -> Text -> IO ()
publish store localPeer =
  mapM_ (\(name, e) -> bind store ("/" <> localPeer <> "/system/type/" <> name) e) coreTypes

File: protocol-generator/julia/src/typedefs.jl
# Core type floor (V8 §9.5) — render-from-model (S4).
#
# The peer publishes its 53 core `system/type/<name>` entities at
# `/{peer}/system/type/{name}`. Each type's data is rendered NATIVELY from an in-code
# declaration here (the single source of truth) through the byte-green S2 codec; the
# resulting content_hash is byte-identical to the Go-rendered
# `type-registry-vectors-v1.cbor` set (the S8 drift target, checked by
# test/typedefs_bytecheck.jl). This is the render-from-model design every peer follows
# (mirrors Zig type_defs.zig / C# CoreTypeRegistry / TS core-type-registry.ts) — the
# profile's "render natively, don't ingest oracle bytes" durable lesson.
#
# Scope is core + operational + type-system bootstrap ONLY — 53 types. Extension
# vocabularies (compute/*, content/*, subscription/*, …) are NOT published by a core
# peer. The oracle's type_system category matches the 53 floor as a hard FAIL gate and
# WARNs (matched-if-present) on the non-floor types it also probes.
#
# Omit-empty semantics: an absent/false field drops the key, so the rendered ECF map is
# byte-identical to the Go reference encoder. The S2 codec sorts map keys canonically
# (RFC 8949 §4.2.1), so declaration order here is irrelevant to the bytes.
module TypeDefs

using ..Cbor: CborMap
using ..Model: Entity, make_entity
using ..Store: ContentStore, store_put!

export publish_types!, all_type_entities, CORE_TYPE_COUNT

# ── field-spec builders (system/type/field-spec shape, omit-empty) ────────────────────
# A field-spec is a CborMap; exactly the set keys present are rendered. Modifiers copy.
_addkey(m::CborMap, k, v) = CborMap(vcat(m.pairs, Pair{Any,Any}[Pair{Any,Any}(k, v)]))

fref(t::AbstractString)            = CborMap(Pair[("type_ref" => String(t))])
opt(s::CborMap)                    = _addkey(s, "optional", true)
sized(s::CborMap, n::Integer)      = _addkey(s, "byte_size", Int(n))
farray(elem::CborMap)              = CborMap(Pair[("array_of" => elem)])
fmap(value::CborMap)               = CborMap(Pair[("map_of" => value)])
fmap(value::CborMap, kt::AbstractString) = CborMap(Pair[("map_of" => value), ("key_type" => String(kt))])
funion(variants::Vector{CborMap})  = CborMap(Pair[("union_of" => Any[v for v in variants])])

# ── type-def builder (system/type entity data, omit-empty) ────────────────────────────
# fields: Vector of (name::String => field-spec::CborMap); layout: Vector{String}.
function type_data(name::AbstractString;
                   extends::Union{Nothing,AbstractString}=nothing,
                   fields=nothing,
                   layout::Union{Nothing,Vector{String}}=nothing)::CborMap
    ps = Pair[("name" => String(name))]
    extends === nothing || push!(ps, "extends" => String(extends))
    if fields !== nothing && !isempty(fields)
        fps = Pair[(String(k) => v) for (k, v) in fields]
        push!(ps, "fields" => CborMap(fps))
    end
    layout === nothing || isempty(layout) || push!(ps, "layout" => Any[s for s in layout])
    return CborMap(ps)
end

type_entity(name; kwargs...) = make_entity("system/type", type_data(name; kwargs...))

# reused nested specs
const sp_string        = fref("primitive/string")
const sp_any           = fref("primitive/any")
const sp_bytes         = fref("primitive/bytes")
const sp_uint          = fref("primitive/uint")
const sp_hash          = fref("system/hash")
const sp_core_entity   = fref("core/entity")
const sp_tree_path     = fref("system/tree/path")
const sp_type_name     = fref("system/type/name")
const sp_grant_entry   = fref("system/capability/grant-entry")
const sp_field_spec    = fref("system/type/field-spec")
const sp_op_spec       = fref("system/handler/operation-spec")
const sp_listing_entry = fref("system/tree/listing-entry")
const sp_type          = fref("system/type")
const sp_multi_granter = fref("system/capability/multi-granter")

# ── the 53 core type definitions (faithful port of the cross-blessed registry) ────────
function all_type_entities()::Vector{Entity}
    T = Entity[]
    push!(T,
        # primitives (8)
        type_entity("primitive/any"),
        type_entity("primitive/bool"),
        type_entity("primitive/bytes"),
        type_entity("primitive/float"),
        type_entity("primitive/int"),
        type_entity("primitive/null"),
        type_entity("primitive/string"),
        type_entity("primitive/uint"),

        # structural roots + envelopes (5)
        type_entity("entity"; fields=[
            "type" => fref("primitive/string"),
            "data" => fref("primitive/any")]),
        type_entity("core/entity"; fields=[
            "type" => fref("primitive/string"),
            "data" => fref("primitive/any"),
            "content_hash" => fref("system/hash")]),
        type_entity("core/envelope"; fields=[
            "root" => fref("core/entity"),
            "included" => opt(fmap(sp_core_entity, "system/hash"))]),
        type_entity("system/envelope"; extends="core/envelope"),
        type_entity("system/protocol/envelope"; extends="core/envelope"),

        # identity / hash / signature (4)
        type_entity("system/hash"; extends="primitive/bytes", fields=[
            "format_code" => sized(fref("primitive/uint"), 1),
            "digest" => fref("primitive/bytes")], layout=["format_code", "digest"]),
        type_entity("system/peer"; fields=[
            "key_type" => fref("primitive/string"),
            "peer_id" => fref("system/peer-id"),
            "public_key" => fref("primitive/bytes")]),
        type_entity("system/peer-id"; extends="primitive/string"),
        type_entity("system/signature"; fields=[
            "algorithm" => fref("primitive/string"),
            "signature" => fref("primitive/bytes"),
            "signer" => fref("system/hash"),
            "target" => fref("system/hash")]),

        # protocol surface (6)
        type_entity("system/protocol/connect/authenticate"; fields=[
            "key_type" => fref("primitive/string"),
            "nonce" => fref("primitive/bytes"),
            "peer_id" => fref("system/peer-id"),
            "public_key" => fref("primitive/bytes")]),
        type_entity("system/protocol/connect/hello"; fields=[
            "protocols" => farray(sp_string),
            "nonce" => fref("primitive/bytes"),
            "peer_id" => fref("system/peer-id"),
            "timestamp" => fref("primitive/uint"),
            "compression" => opt(farray(sp_string)),
            "encryption" => opt(farray(sp_string)),
            "hash_formats" => opt(farray(sp_string)),
            "key_types" => opt(farray(sp_string))]),
        type_entity("system/protocol/error"; fields=[
            "code" => fref("primitive/string"),
            "message" => opt(fref("primitive/string")),
            "rejected_marker" => opt(fref("system/hash"))]),
        type_entity("system/protocol/execute"; fields=[
            "operation" => fref("primitive/string"),
            "params" => fref("core/entity"),
            "request_id" => fref("primitive/string"),
            "uri" => fref("system/tree/path"),
            "author" => opt(fref("system/hash")),
            "bounds" => opt(fref("system/bounds")),
            "capability" => opt(fref("system/hash")),
            "deliver_to" => opt(fref("system/delivery-spec")),
            "deliver_token" => opt(fref("system/hash")),
            "durability_request" => opt(fref("system/durability-request")),
            "resource" => opt(fref("system/protocol/resource-target"))]),
        type_entity("system/protocol/execute/response"; fields=[
            "request_id" => fref("primitive/string"),
            "result" => fref("core/entity"),
            "status" => fref("primitive/uint"),
            "durability" => opt(fref("system/durability-result"))]),
        type_entity("system/protocol/resource-target"; fields=[
            "targets" => farray(sp_tree_path),
            "exclude" => opt(farray(sp_tree_path))]),

        # capability (12)
        type_entity("system/capability/grant"; fields=[
            "token" => fref("system/hash")]),
        type_entity("system/capability/grant-entry"; fields=[
            "handlers" => fref("system/capability/path-scope"),
            "operations" => fref("system/capability/id-scope"),
            "resources" => fref("system/capability/path-scope"),
            "allowances" => opt(fmap(sp_any)),
            "constraints" => opt(fmap(sp_any)),
            "peers" => opt(fref("system/capability/id-scope"))]),
        type_entity("system/capability/id-scope"; fields=[
            "include" => farray(sp_string),
            "exclude" => opt(farray(sp_string))]),
        type_entity("system/capability/path-scope"; fields=[
            "include" => farray(sp_tree_path),
            "exclude" => opt(farray(sp_tree_path))]),
        type_entity("system/capability/request"; fields=[
            "grants" => farray(sp_grant_entry),
            "ttl_ms" => opt(fref("primitive/uint"))]),
        type_entity("system/capability/revocation"; fields=[
            "token" => fref("system/hash"),
            "revoked_at" => fref("primitive/uint"),
            "reason" => opt(fref("primitive/string"))]),
        type_entity("system/capability/revoke-request"; fields=[
            "token" => fref("system/hash"),
            "reason" => opt(fref("primitive/string"))]),
        type_entity("system/capability/delegate-request"; fields=[
            "grants" => farray(sp_grant_entry),
            "parent" => fref("system/hash"),
            "ttl_ms" => opt(fref("primitive/uint"))]),
        type_entity("system/capability/delegation-caveats"; fields=[
            "max_delegation_depth" => opt(fref("primitive/uint")),
            "max_delegation_ttl" => opt(fref("primitive/uint")),
            "no_delegation" => opt(fref("primitive/bool"))]),
        type_entity("system/capability/policy-entry"; fields=[
            "grants" => farray(sp_grant_entry),
            "peer_pattern" => fref("primitive/string"),
            "notes" => opt(fref("primitive/string")),
            "ttl_ms" => opt(fref("primitive/uint"))]),
        type_entity("system/capability/token"; fields=[
            "created_at" => fref("primitive/uint"),
            "grantee" => fref("system/hash"),
            "granter" => funion(CborMap[sp_hash, sp_multi_granter]),
            "grants" => farray(sp_grant_entry),
            "delegation_caveats" => opt(fref("system/capability/delegation-caveats")),
            "expires_at" => opt(fref("primitive/uint")),
            "not_before" => opt(fref("primitive/uint")),
            "parent" => opt(fref("system/hash")),
            "resource_limits" => opt(fref("system/resource-limits"))]),
        type_entity("system/capability/multi-granter"; fields=[
            "signers" => farray(sp_hash),
            "threshold" => fref("primitive/uint")]),

        # handler machinery (6)
        type_entity("system/handler"; fields=[
            "interface" => fref("system/tree/path"),
            "expression_path" => opt(fref("system/tree/path")),
            "internal_scope" => opt(farray(sp_grant_entry)),
            "max_scope" => opt(farray(sp_grant_entry))]),
        type_entity("system/handler/interface"; fields=[
            "name" => fref("primitive/string"),
            "operations" => fmap(sp_op_spec),
            "pattern" => fref("system/tree/path")]),
        type_entity("system/handler/manifest"; extends="system/handler/interface", fields=[
            "name" => fref("primitive/string"),
            "operations" => fmap(sp_op_spec),
            "pattern" => fref("system/tree/path"),
            "expression_path" => opt(fref("system/tree/path")),
            "internal_scope" => opt(farray(sp_grant_entry)),
            "max_scope" => opt(farray(sp_grant_entry))]),
        type_entity("system/handler/operation-spec"; fields=[
            "input_type" => opt(fref("system/type/name")),
            "output_type" => opt(fref("system/type/name"))]),
        type_entity("system/handler/register-request"; fields=[
            "manifest" => fref("system/handler/manifest"),
            "requested_scope" => opt(farray(sp_grant_entry)),
            "types" => opt(fmap(sp_type))]),
        type_entity("system/handler/register-result"; fields=[
            "grant" => fref("system/capability/token"),
            "pattern" => fref("system/tree/path")]),

        # tree (5)
        type_entity("system/tree/get-request"; fields=[
            "limit" => opt(fref("primitive/uint")),
            "mode" => opt(fref("primitive/string")),
            "offset" => opt(fref("primitive/uint")),
            "tree_id" => opt(fref("primitive/string"))]),
        type_entity("system/tree/put-request"; fields=[
            "entity" => opt(fref("core/entity")),
            "expected_hash" => opt(fref("system/hash")),
            "tree_id" => opt(fref("primitive/string"))]),
        type_entity("system/tree/listing"; fields=[
            "count" => fref("primitive/uint"),
            "entries" => fmap(sp_listing_entry),
            "offset" => fref("primitive/uint"),
            "path" => fref("system/tree/path"),
            "next_page" => opt(fref("system/hash"))]),
        type_entity("system/tree/listing-entry"; fields=[
            "has_children" => fref("primitive/bool"),
            "hash" => opt(fref("system/hash"))]),
        type_entity("system/tree/path"; extends="primitive/string"),

        # type-system bootstrap (3)
        type_entity("system/type"; fields=[
            "name" => fref("system/type/name"),
            "extends" => opt(fref("system/type/name")),
            "fields" => opt(fmap(sp_field_spec)),
            "layout" => opt(farray(sp_string)),
            "type_args" => opt(fmap(sp_type_name)),
            "type_params" => opt(farray(sp_string))]),
        type_entity("system/type/field-spec"; fields=[
            "type_ref" => opt(fref("system/type/name")),
            "optional" => opt(fref("primitive/bool")),
            "array_of" => opt(fref("system/type/field-spec")),
            "map_of" => opt(fref("system/type/field-spec")),
            "union_of" => opt(farray(sp_field_spec)),
            "key_type" => opt(fref("system/type/name")),
            "byte_size" => opt(fref("primitive/uint")),
            "type_param" => opt(fref("primitive/string")),
            "type_args" => opt(fmap(sp_type_name)),
            "default" => opt(fref("primitive/any")),
            "constraints" => opt(farray(sp_core_entity))]),
        type_entity("system/type/name"; extends="primitive/string"),

        # operational (4)
        type_entity("system/bounds"; fields=[
            "budget" => opt(fref("primitive/uint")),
            "cascade_depth" => opt(fref("primitive/uint")),
            "chain_id" => opt(fref("primitive/string")),
            "parent_chain_id" => opt(fref("primitive/string")),
            "ttl" => opt(fref("primitive/uint")),
            "visited" => opt(farray(sp_tree_path))]),
        type_entity("system/resource-limits"; fields=[
            "max_budget" => opt(fref("primitive/uint")),
            "max_ttl" => opt(fref("primitive/uint")),
            "max_visited_length" => opt(fref("primitive/uint"))]),
        type_entity("system/delivery-spec"; fields=[
            "operation" => fref("primitive/string"),
            "uri" => fref("system/tree/path")]),
        type_entity("system/deletion-marker"),
    )
    return T
end

const CORE_TYPE_COUNT = 53

"""Seed every core type entity into the store at `/{peer}/system/type/<name>`."""
function publish_types!(st::ContentStore, local_peer::AbstractString)
    for e in all_type_entities()
        name = e.data.pairs[findfirst(p -> p.first == "name", e.data.pairs)].second
        store_put!(st, e; path="/$(local_peer)/system/type/$(name)")
    end
    return nothing
end

end # module TypeDefs
